src/data_structures/seg_tree.rs

```diff
@@ -1,135 +1,94 @@
 //! # Segment Tree
 
-use std::ops::Range;
-
-/// see https://codeforces.com/blog/entry/112755
-///
-/// returns the split point of the range which
-/// makes the segment tree a complete binary tree
-fn split(tr: &Range<usize>) -> usize {
-    let pw2 = 1 << tr.len().ilog2();
-    (tr.start + pw2).min(tr.end - pw2 / 2)
-}
-
-fn op(vl: u64, vr: u64) -> u64 {
-    vl + vr
-}
-
 /// # Example
 /// ```
 /// use programming_team_code_rust::data_structures::seg_tree::SegTree;
 ///
-/// let mut seg_tree = SegTree::new(10);
-/// seg_tree.update(0..5, 1);
-/// seg_tree.update(5..10, 2);
-/// assert_eq!(seg_tree.query(0..10), 15);
+/// let md = 6;
+/// let mut st = SegTree::<usize>::new(3, move |x, y| (x + y) % md, 0);
+/// st.set(1, 2);
+/// st.set(2, 3);
+/// assert_eq!(st.query(0..3), 5);
 /// ```
-pub struct SegTree {
+pub struct SegTree<T> {
     n: usize,
-    tree: Vec<u64>,
-    lazy: Vec<u64>,
+    op: Box<dyn Fn(T, T) -> T>,
+    unit: T,
+    tree: Vec<T>,
 }
 
-impl SegTree {
+impl<T: Clone> SegTree<T> {
     /// Creates a new segment tree with n elements
     ///
     /// # Complexity
     /// - Time: O(n)
     /// - Space: O(n)
-    pub fn new(n: usize) -> Self {
-        SegTree {
+    pub fn new(n: usize, op: impl Fn(T, T) -> T + 'static, unit: T) -> Self {
+        Self {
             n,
-            tree: vec![0; 2 * n],
-            lazy: vec![0; n],
+            op: Box::new(op),
+            unit: unit.clone(),
+            tree: vec![unit.clone(); 2 * n],
         }
     }
 
-    /// Constructs a segment tree on a given array
+    /// Creates a segment tree on a given array
     ///
     /// # Complexity
     /// - Time: O(n)
     /// - Space: O(n)
-    pub fn build_on_array(a: &[u64]) -> Self {
+    pub fn build_on_array(a: &[T], op: impl Fn(T, T) -> T + 'static, unit: T) -> Self {
         let n = a.len();
-        let mut pw2 = 1;
-        while pw2 < n {
-            pw2 *= 2;
-        }
-        let mut tree = vec![0; 2 * n];
-        for i in 0..n {
-            tree[(i + pw2) % n + n] = a[i];
-        }
+        let mut tree = vec![unit.clone(); n];
+        tree.extend(a.to_vec());
         for i in (1..n).rev() {
-            tree[i] = op(tree[2 * i], tree[2 * i + 1]);
+            tree[i] = op(tree[2 * i].clone(), tree[2 * i + 1].clone());
```

Comment:
should we maybe have the `op` take the items by reference?

Reply:
good catch